contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,168 @@
+            // With current approach, even if one action fails, `data` will be empty bytes even for successful
+            // actions.
+            if (!isActionSuccessful) {
+                revert ActionExecutionFailed(address(safe), transaction.metaHash, i);
+            } else {
+                emit ActionExecuted(address(safe), transaction.metaHash, i);
+                data[i] = resultData;
+            }
+        }
+    }
+
+    /**
+     * @notice TODO
+     * @param safe TODO
+     * @param rootAccess TODO
+     * @return data TODO
+     */
+    function executeRootAccess(
+        ISafe safe,
+        SafeRootAccess calldata rootAccess
+    ) external override onlyEnabledModule(safe) returns (bytes memory data) {
+        SafeProtocolAction memory safeProtocolAction = rootAccess.action;
+        // TODO: Check for re-entrancy attacks
+        // TODO: Validate metahash
+
+        // Re-confirm if this check if needed and correct.
+        if (!ISafeProtocolModule(msg.sender).requiresRootAccess()) {
+            revert ModuleRequiresRootAccess(msg.sender);
+        }
+
+        if (!enabledComponents[address(safe)][msg.sender].rootAddressGranted) {
+            // TODO: Need new error type?
+            revert ModuleRequiresRootAccess(msg.sender);
+        }
+        bool success;
+        (success, data) = safe.execTransactionFromModuleReturnData(
+            safeProtocolAction.to,
+            safeProtocolAction.value,
+            safeProtocolAction.data,
+            1
+        );
+        if (success) {
+            emit RootAccessActionExecuted(address(safe), rootAccess.metaHash);
+        } else {
+            revert RootAccessActionExecutionFailed(address(safe), rootAccess.metaHash);
+        }
+    }
+
+    /**
+     * @notice Called by a Safe to enable a module on a Safe. To be called by a safe.
+     * @param module ISafeProtocolModule A module that has to be enabled
+     * @param allowRootAccess Bool indicating whether root access to be allowed.
+     */
+    function enableModule(ISafeProtocolModule module, bool allowRootAccess) external {
+        // TODO: Check if module is a valid address and implements valid interface.
+        //       Validate if caller is a Safe.
+        //       Should it be allowed to enable a module twice with different allowRootAccess flag?
+        if (allowRootAccess != module.requiresRootAccess()) {
+            revert ModuleAccessMismatch(address(module), module.requiresRootAccess(), allowRootAccess);
+        }
+        enabledComponents[msg.sender][address(module)] = MoudleAccessInfo(true, allowRootAccess);
+
+        emit ModuleEnabled(msg.sender, address(module), allowRootAccess);
+    }
+
+    /**
+     * @notice Disable a module. This function should be called by Safe.
+     * @param module Module to be disabled
+     */
+    function disableModule(ISafeProtocolModule module) external {
+        // TODO: Validate if caller is a Safe
```

Comment:
What attack vectors do we see if this is not a Safe?

Reply:
let me rephrase here, if module is enabled on a safe, it send a transaction through the safe that disables other module (which ideally should not happen). I put the comment here initially as a remainder to enforce the flow of transaction as below:

Safe owner -> Safe -> Mediator `disableModule`.

One possible way is to add nonReentrant modifier to all 4 functions: executeTransaction, executeRootAccess, enableModule, disableModule

Reply:
See above